controller/src/main/java/io/pravega/controller/server/bucket/ZooKeeperBucketManager.java

```diff
@@ -62,69 +64,95 @@ protected int getBucketCount() {
-                pathChildrenCache.clear();
-                pathChildrenCache.close();
-            } catch (IOException e) {
-                log.warn("unable to close listener for bucket ownership", e);
+    public void addBucketControllerMapListener() {
+        ZKUtils.createPathIfNotExists(bucketStore.getClient(), bucketStore.getBucketControllerMapPath(getServiceType()),
+                BucketControllerMap.EMPTY.toBytes());
+        NodeCache cache = bucketStore.getBucketControllerMapNodeCache(getServiceType());
+        cache.getListenable().addListener(this::handleBuckets);
+        log.info("{}: Bucket controller map listener registered.", getServiceType());
+        cache.start(true);
+        manageBuckets(cluster.getClusterMembers().size()).whenComplete((r, e) -> {
+            if (e == null) {
+                log.debug("{}: Manage buckets completes with result: {}.", getServiceType(), r);
+            } else {
+                log.error("{}: Manage buckets completes with result: {} and exception: {}.", getServiceType(), r, e);
             }
-        }
+        });
     }
 
     @Override
-    public CompletableFuture<Void> initializeService() {
-        return bucketStore.createBucketsRoot(getServiceType());
+    public void startLeaderElection() {
+        String bucketDistributorLeader = "bucketDistributorLeader";
+        String leaderSubPath = ZKPaths.makePath("cluster", getServiceType().getName());
+        String leaderZKPath = ZKPaths.makePath(leaderSubPath, bucketDistributorLeader);
+
+        leaderSelector = new LeaderSelector(bucketStore.getClient(), leaderZKPath, bucketManagerLeader);
+        //Listen for any zookeeper connection state changes
+        bucketStore.getClient().getConnectionStateListenable().addListener(
+                (curatorClient, newState) -> {
+                    switch (newState) {
+                        case LOST:
+                            log.warn("Connection to zookeeper lost, attempting to interrrupt the leader thread");
+                            leaderSelector.interruptLeadership();
+                            break;
+                        case SUSPENDED:
+                            if (leaderSelector.hasLeadership()) {
+                                log.info("Zookeeper session suspended, pausing the bucket manager");
+                                bucketManagerLeader.suspend();
+                            }
+                            break;
+                        case RECONNECTED:
+                            if (leaderSelector.hasLeadership()) {
+                                log.info("Zookeeper session reconnected, resume the bucket manager");
+                                bucketManagerLeader.resume();
+                            }
+                            break;
+                        //$CASES-OMITTED$
+                        default:
+                            log.debug("Connection state to zookeeper updated: " + newState.toString());
```

Comment:
I think it is worth having this log line at least in `info` level, as otherwise we do not realize if unknown events are getting processed by this class. And always use `{}`, not string concatenation.

Reply:
Done.